packages/webapp/src/containers/Animals/AddAnimals/AddAnimalDetails/index.tsx

```diff
@@ -13,8 +13,172 @@
+    organicStatusOptions,
+    originOptions,
+  } = useAnimalOptions(
+    'default_types',
+    'sex',
+    'sexDetails',
+    'use',
+    'tagType',
+    'tagColor',
+    'organicStatus',
+    'origin',
+  );
+
+  /* Populate details form based on basics data */
+  usePopulateDetails(replace);
+
+  /* Render logic */
+  const currency = useCurrencySymbol();
+
+  const isAnimalField = (field: AnimalDetailsField) =>
+    field.animal_or_batch === AnimalOrBatchKeys.ANIMAL;
+
+  const generateExpandableItem = (field: AnimalDetailsField, index: number) => {
+    const isAnimal = isAnimalField(field);
+
+    const fieldArrayIndex = fields.findIndex(({ id }) => id === field.id);
+    const namePrefix = `${STEPS.DETAILS}.${fieldArrayIndex}.` as const;
+
+    const isExpanded = expandedIds.includes(field.id);
+
+    const watchedCount = watch(`${namePrefix}${DetailsFields.COUNT}`);
+
+    const useOptionsForType = useOptions.find(
+      ({ default_type_id }: { default_type_id: string | null }) =>
+        default_type_id === `${parseUniqueDefaultId(field.type.value)}` || default_type_id === null,
+    );
+
+    const mainContent = (
+      <AnimalFormHeaderItem
+        showRemove={fields.length > 1}
+        key={field.id}
+        isExpanded={isExpanded}
+        onRemove={() => onRemoveCard(fieldArrayIndex)}
+        type={field.type.label}
+        breed={field.breed?.label}
+        totalCount={isAnimal ? animalCount : batchCount}
+        number={index + 1}
+        iconKey={getDefaultAnimalIconName(defaultTypes, parseUniqueDefaultId(field.type.value))}
+        isBatch={!isAnimal}
+        count={!isAnimal ? watchedCount : undefined}
+        sex={sexOptions.label}
+      />
+    );
+
+    const commonProps = {
+      key: field.id,
+      generalDetailProps: {
+        sexOptions,
+        sexDetailsOptions,
+        useOptions: useOptionsForType.uses,
+      },
+      otherDetailsProps: { organicStatusOptions },
+      originProps: {
+        currency: currency,
+        originOptions,
+      },
+      namePrefix: namePrefix,
+    };
+
+    const expandedContent = isAnimal ? (
+      <AnimalDetails {...commonProps} uniqueDetailsProps={{ tagTypeOptions, tagColorOptions }} />
+    ) : (
+      <BatchDetails {...commonProps} />
+    );
+
+    return (
+      <div key={field.id}>
+        <ExpandableItem
+          itemKey={field.id}
+          isExpanded={isExpanded}
+          iconClickOnly={false}
+          onClick={() => toggleExpanded(field.id)}
+          mainContent={mainContent}
+          expandedContent={expandedContent}
+          classes={{
+            expandedContainer: styles.expandedContainer,
+            mainContentWithIcon: styles.expandableHeader,
+            expandedMainContentWithIcon: styles.expandedHeader,
+          }}
+          leftCollapseIcon
+        />
+      </div>
+    );
+  };
+
+  const animals = fields.filter(isAnimalField);
+  const batches = fields.filter((field) => !isAnimalField(field));
```

Comment:
I'm hoping to avoid finding `fieldArrayIndex` in `generateExpandableItem`.

Is it safe to assume that the indices of the `fields` array correspond directly to `fieldArrayIndex`? If so, we could create `animals` and `batches` arrays while attaching `fieldArrayIndex`:

```
const animals = [] as (AnimalDetailsField & { fieldArrayIndex: number })[];
const batches = [] as (AnimalDetailsField & { fieldArrayIndex: number })[];

fields.forEach((animalOrBatch, index) => {
  const array = isAnimalField(animalOrBatch) ? animals : batches;
  array.push({ ...animalOrBatch, fieldArrayIndex: index });
});
```
Then, use it in the function: 

```
const generateExpandableItem = (
  fieldWithFieldArrayIndex: AnimalDetailsField & { fieldArrayIndex: number },
  index: number,
) => {
  const { fieldArrayIndex, ...field } = fieldWithFieldArrayIndex;
...
``` 
What do you think?

Reply:
Well, it's not my first choice for the same reason I went with the double filter instead of `fields.forEach()` -- I usually prefer a little more readability over a little more efficiency for these small arrays -- but now I might be letting the inefficient things add up, so I'll use it 🙏 

(Also I like this a lot: `const array = isAnimalField(animalOrBatch) ? animals : batches` 😄)